File: tal_algo/private/tiling_1x2-bars_in_mxn-grids/turingarena-files/C/skeleton.c
#include <stdio.h>
#include <stdlib.h>

static const int H = 0;

static const int V = 1;


int is_tilable(int m, int n);

void compose_tiling(int m, int n, void place_tile(int row, int col, int dir));


int main() {
    // checkpoint
    printf("%d\n", 0);
    // read m, n
    static int n;
    static int m;
    fflush(stdout);
    scanf("%d%d", &m, &n);
    // call res = is_tilable(m, n)
    static int res;
    res = is_tilable(m, n);
    // write res
    printf("%d\n", res);
    // if res {...}
    if (res) {
        // read choice
        static int choice;
        fflush(stdout);
        scanf("%d", &choice);
        // if choice {...}
        if (choice) {
            // read m1, n1
            static int m1;
            static int n1;
            fflush(stdout);
            scanf("%d%d", &m1, &n1);
            // call compose_tiling(m1, n1) callbacks {...}
            {
                void place_tile(int row, int col, int dir) {
                    // callback place_tile
                    printf("%d %d\n", 1, 0);
                    // write row, col, dir
                    printf("%d %d %d\n", row, col, dir);
                }
                compose_tiling(m1, n1, place_tile);
            }
            // no more callbacks
            printf("%d %d\n", 0, 0);
        }
    }
    // exit
    exit(0);
}

File: tal_algo/private/tiling_1x2-bars_in_mxn-grids/turingarena-files/Rust/skeleton.rs
mod solution;
use std::io::Write;

macro_rules! readln {
    ($($var:expr),*) => {{
        let mut buf = String::new();
        std::io::stdin().read_line(&mut buf).unwrap();
        let parts: Vec<&str> = buf.trim().split(" ").collect();
        let mut i: usize = 0;
        $(
            assert!(i < parts.len(), "input format incorrect: too few values on this line"); 
            $var = parts[i].parse().unwrap();
            i += 1;
        )*
        assert!(i == parts.len(), "input format incorrect: too many values on the line")
    }};
}

const H: i64 = 0;
const V: i64 = 1;

fn main() {
    // checkpoint
    println!("{}", 0);
    // read m, n
    let n: i64;
    let m: i64;
    std::io::stdout().flush().unwrap();
    readln!(m, n);
    // call res = is_tilable(m, n)
    let res: i64;
    res = solution::is_tilable(m, n);
    // write res
    println!("{}", res);
    // if res {...}
    if res != 0 {
        // read choice
        let choice: i64;
        std::io::stdout().flush().unwrap();
        readln!(choice);
        // if choice {...}
        if choice != 0 {
            // read m1, n1
            let m1: i64;
            let n1: i64;
            std::io::stdout().flush().unwrap();
            readln!(m1, n1);
            // call compose_tiling(m1, n1) callbacks {...}
            solution::compose_tiling(m1, n1, |row: i64, col: i64, dir: i64| {
                // callback place_tile
                println!("{} {}", 1, 0);
                // write row, col, dir
                println!("{} {} {}", row, col, dir);
            });
            // no more callbacks
            println!("{} {}", 0, 0);
        }
    }
    // exit
    std::process::exit(0);
}
